org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/typeHierarchy/TypeHierarchyView.java

```diff
@@ -74,21 +385,99 @@ private void updateHierarchyModeItem(ToolItem hierchyModeItem, boolean showSuper
 				hierchyModeItem.setToolTipText(showSuperTypes ? Messages.typeHierarchy_show_subtypes : Messages.typeHierarchy_show_supertypes);
 			}
 		};
-		treeViewer = filteredTree.getViewer();
-		treeViewer.setContentProvider(contentProvider);
+	}
 
-		treeViewer.setLabelProvider(new TypeHierarchyItemLabelProvider());
+	private SymbolsContainer getSymbolsContainer(IFile file) {
+		SymbolsContainer symbolsContainer = cachedSymbols.get(file.getLocationURI());
+		if (symbolsContainer == null) {
+			symbolsContainer = new SymbolsContainer(file);
+			cachedSymbols.put(file.getLocationURI(), symbolsContainer);
+		}
+		return symbolsContainer;
+	}
 
-		treeViewer.setUseHashlookup(true);
-		treeViewer.getControl().setEnabled(false);
-		treeViewer.addDoubleClickListener(new IDoubleClickListener() {
-			@Override
-			public void doubleClick(final DoubleClickEvent event) {
-				TypeHierarchyItem item = (TypeHierarchyItem)((IStructuredSelection)event.getSelection()).getFirstElement();
-				LSPEclipseUtils.open(item.getUri(), item.getSelectionRange());
+	private void refreshSymbols(SymbolsContainer symbolsContainer, boolean documentModified) {
+		if (symbolsContainer == null || (!symbolsContainer.isDirty && !documentModified)) {
+			return;
+		}
+		final IDocument document = symbolsContainer.getDocument();
+		try {
+			if (document != null) {
+				final var params = new DocumentSymbolParams(
+						LSPEclipseUtils.toTextDocumentIdentifier(document));
+				CompletableFuture<Optional<LanguageServerWrapper>> languageServer = LanguageServers
+						.forDocument(document)
+						.withFilter(
+								capabilities -> LSPEclipseUtils.hasCapability(capabilities.getDocumentSymbolProvider()))
+						.computeFirst((w, ls) -> CompletableFuture.completedFuture(w));
+				try {
+					symbols = languageServer.get(500, TimeUnit.MILLISECONDS).filter(Objects::nonNull)
```

Comment:
if symbols contains already a future, should we not cancel it?

Reply:
The calling method `refreshSymbols` returns when the last completions stage has joined (see line  436). `refreshSymbols` will be called by the `synchronized` method `refreshMemberViewer`. That's why I assumed that the future cannot be called again until it has been completed.

Reply:
Correct me if I am wrong